File: 493. Reverse Pairs/main.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    int Merge(vector<int>& nums, vector<int>& tmp, int l, int m, int r) {
        int i = l, j = m, k = l, res = 0;
        while (i < m && j < r) {
            if (nums[i] > 2LL * nums[j]) {
                res += m - i;
                ++j;
            } else {
                ++i;
            }
        }
        i = l, j = m;
        while (i < m && j < r) {
            if (nums[i] > nums[j]) {
                tmp[k++] = nums[j++];
            } else {
                tmp[k++] = nums[i++];
            }
        }
        int n = i < m ? i : j;
        while (k < r) tmp[k++] = nums[n++];
        for (int i = l; i < r; ++i) nums[i] = tmp[i];
        return res;
    }
    int MergeSort(vector<int>& nums, vector<int>& tmp, int l, int r) {
        if (r - l <= 1) return 0;
        int m = (l + r) / 2, res = 0;
        res += MergeSort(nums, tmp, l, m);
        res += MergeSort(nums, tmp, m, r);
        res += Merge(nums, tmp, l, m, r);
        return res;
    }
    int reversePairs(vector<int>& nums) {
        vector<int> tmp(nums.size());
        return MergeSort(nums, tmp, 0, nums.size());
    }
};

int main()
{
    Solution sol;
    int T; cin >> T;
    while (T--) {
        int n; cin >> n;
        vector<int> vec(n);
        for (int i = 0; i != n; ++i)
            cin >> vec[i];
        cout << sol.reversePairs(vec) << endl;
    }
    return 0;
}
/*
Input:
2
5
1 3 2 3 1
5
2 4 3 5 1
Output:
2
3
*/

File: 493. Reverse Pairs/main.rs
#![allow(dead_code)]
struct Solution;
fn merge(arr: &mut Vec<i32>, tmp: &mut Vec<i32>, l: usize, m: usize, r: usize) -> i32 {
    let mut res = 0;
    let (mut i, mut j) = (l, m);
    while i < m && j < r {
        if arr[i] as i64 <= 2 * arr[j] as i64 {
            i += 1;
        } else {
            res += m - i;
            j += 1;
        }
    }
    let (mut i, mut j, mut k) = (l, m, l);
    while i < m && j < r {
        if arr[i] < arr[j] {
            tmp[k] = arr[i];
            i += 1;
        } else {
            tmp[k] = arr[j];
            j += 1;
        }
        k += 1;
    }
    let mut t = if i < m { i } else { j };
    while k < r {
        tmp[k] = arr[t];
        t += 1;
        k += 1;
    }
    for i in l..r {
        arr[i] = tmp[i];
    }
    res as i32
}
fn merge_sort(arr: &mut Vec<i32>, tmp: &mut Vec<i32>, l: usize, r: usize) -> i32 {
    if r - l <= 1 {
        return 0;
    }
    let (mut res, m) = (0, (l + r) / 2);
    res += merge_sort(arr, tmp, l, m);
    res += merge_sort(arr, tmp, m, r);
    res += merge(arr, tmp, l, m, r);
    res
}
impl Solution {
    pub fn reverse_pairs(mut nums: Vec<i32>) -> i32 {
        let n = nums.len();
        merge_sort(&mut nums, &mut vec![0; n], 0, n)
    }
}

fn main() {}
